src/reseeding.rs

```diff
@@ -39,73 +40,127 @@ use {RngCore, SeedableRng, Error, ErrorKind};
+        self.0.next_u64()
+    }
+
+    fn fill_bytes(&mut self, dest: &mut [u8]) {
+        self.0.fill_bytes(dest)
+    }
+
+    fn try_fill_bytes(&mut self, dest: &mut [u8]) -> Result<(), Error> {
+        self.0.try_fill_bytes(dest)
+    }
+}
+
+#[derive(Debug)]
+struct Reseeder<R, Rsdr> {
+    core: R,
+    reseeder: Rsdr,
+    threshold: i64,
+    bytes_until_reseed: i64,
+}
+
+impl<R, Rsdr> BlockRngCore<u32> for Reseeder<R, Rsdr>
+where R: BlockRngCore<u32> + SeedableRng,
+      Rsdr: RngCore
+{
+    type Results = <R as BlockRngCore<u32>>::Results;
+
+    fn generate(&mut self, results: &mut Self::Results) -> Result<(), Error> {
+        if self.bytes_until_reseed <= 0 {
+            // We want to reseed here, and generate results later in the
+            // function. If generating results fail, we should return the error
+            // from that. If generating results succeeded, but reseeding failed,
+            // we should return the error from reseeding.
+            // The only way to get this behaviour without destroying performance
+            // was to split part of the function out into a
+            // `reseed_and_generate` method.
+            return self.reseed_and_generate(results);
         }
+        self.bytes_until_reseed -= results.as_ref().len() as i64 * 4;
+        self.core.generate(results)
     }
+}
 
+impl<R, Rsdr> Reseeder<R, Rsdr>
+where R: BlockRngCore<u32> + SeedableRng,
+      Rsdr: RngCore
+{
     /// Reseed the internal PRNG.
-    ///
-    /// This will try to work around errors in the RNG used for reseeding
-    /// intelligently through some combination of retrying and delaying
-    /// reseeding until later. So long as the internal PRNG doesn't fail, this
-    /// method will not fail, i.e. failures from the reseeding source are not
-    /// fatal.
-    pub fn reseed(&mut self) {
-        // Behaviour is identical to `try_reseed`; we just squelch the error.
-        let _res = self.try_reseed();
+    fn reseed(&mut self) -> Result<(), Error> {
+        R::from_rng(&mut self.reseeder).map(|result| self.core = result)
     }
 
-    /// Reseed the internal RNG if the number of bytes that have been
-    /// generated exceed the threshold.
+    /// Reseed the internal PRNG.
     ///
-    /// If reseeding fails, return an error with the original cause. Note that
-    /// in case of error we simply delay reseeding, allowing the generator to
-    /// continue its output of random data and try reseeding again later;
-    /// because of this we always return kind `ErrorKind::Transient`.
-    #[inline(never)]
-    pub fn try_reseed(&mut self) -> Result<(), Error> {
+    /// If reseeding fails, this will try to work around errors intelligently
+    /// through some combination of retrying and delaying reseeding until later.
+    /// It will also report the error with `ErrorKind::Transient` with the
+    /// original error as cause.
+    fn auto_reseed(&mut self) -> Result<(), Error> {
```

Comment:
I don't really like the changes to the name and doc. The name `auto_reseed` doesn't really imply anything more than `reseed` (since obviously it's talking about itself).